controllers/clusters/on_premises.go

```diff
@@ -0,0 +1,709 @@
+				PersistentVolumeClaim: &virtcorev1.PersistentVolumeClaimVolumeSource{
+					PersistentVolumeClaimVolumeSource: k8scorev1.PersistentVolumeClaimVolumeSource{
+						ClaimName: dvName,
+					},
+				},
+			},
+		})
+	}
+
+	return vm, nil
+}
+
+func newExposeService(
+	b *onPremiseBootstrap,
+	svcName,
+	vmName,
+	nodeID string,
+) *k8scorev1.Service {
+	return &k8scorev1.Service{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       models.ServiceKind,
+			APIVersion: models.K8sAPIVersionV1,
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      svcName,
+			Namespace: b.K8sObject.GetNamespace(),
+			Labels: map[string]string{
+				models.ClusterIDLabel: b.ClusterStatus.ID,
+				models.NodeIDLabel:    nodeID,
+			},
+			Finalizers: []string{models.DeletionFinalizer},
+		},
+		Spec: k8scorev1.ServiceSpec{
+			Ports: b.ExposePorts,
+			Selector: map[string]string{
+				models.KubevirtDomainLabel: vmName,
+				models.NodeIDLabel:         nodeID,
+			},
+			Type: models.LBType,
+		},
+	}
+}
+
+func newHeadlessService(
+	b *onPremiseBootstrap,
+	svcName string,
+) *k8scorev1.Service {
+	return &k8scorev1.Service{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       models.ServiceKind,
+			APIVersion: models.K8sAPIVersionV1,
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      svcName,
+			Namespace: b.K8sObject.GetNamespace(),
+			Labels: map[string]string{
+				models.ClusterIDLabel: b.ClusterStatus.ID,
+			},
+			Finalizers: []string{models.DeletionFinalizer},
+		},
+		Spec: k8scorev1.ServiceSpec{
+			ClusterIP: "None",
+			Ports:     b.HeadlessPorts,
+			Selector: map[string]string{
+				models.ClusterIDLabel: b.ClusterStatus.ID,
+				models.NodeLabel:      models.WorkerNode,
+			},
+		},
+	}
+}
+
+func deleteOnPremResources(ctx context.Context, K8sClient client.Client, clusterID, ns string) error {
+	vms := &virtcorev1.VirtualMachineList{}
+	err := K8sClient.List(ctx, vms, &client.ListOptions{
+		LabelSelector: labels.SelectorFromSet(map[string]string{
+			models.ClusterIDLabel: clusterID,
+		}),
+		Namespace: ns,
+	})
+	if err != nil {
+		return err
+	}
+
+	for _, vm := range vms.Items {
+		err = K8sClient.Delete(ctx, &vm)
+		if err != nil {
+			return err
+		}
+
+		patch := client.MergeFrom(vm.DeepCopy())
+		controllerutil.RemoveFinalizer(&vm, models.DeletionFinalizer)
+		err = K8sClient.Patch(ctx, &vm, patch)
+		if err != nil {
+			return err
+		}
+	}
+
+	vmis := &virtcorev1.VirtualMachineInstanceList{}
+	err = K8sClient.List(ctx, vmis, &client.ListOptions{
+		LabelSelector: labels.SelectorFromSet(map[string]string{
+			models.ClusterIDLabel: clusterID,
+		}),
+		Namespace: ns,
+	})
+	if err != nil {
+		return err
+	}
+
+	for _, vmi := range vmis.Items {
+		err = K8sClient.Delete(ctx, &vmi)
+		if err != nil {
+			return err
+		}
+
+		patch := client.MergeFrom(vmi.DeepCopy())
+		controllerutil.RemoveFinalizer(&vmi, models.DeletionFinalizer)
+		err = K8sClient.Patch(ctx, &vmi, patch)
+		if err != nil {
+			return err
+		}
+	}
```

Comment:
VMIs are deleted when their VM resource is deleted isn't?
If so we don't need to explicitly delete VMIs

Reply:
Let's be totally sure that all resources are deleted